File: bin/rough_heston_gatheral/price_european.R
#!/usr/bin/env Rscript
# Rough Heston pricer wrapper (jgatheral/RationalRoughHeston, Adams scheme)
# Issue #1200 interface:
#   price_european.R S0 K T H lambda theta nu rho r
# Prints a single line:  call_price put_price
# Convention (Gatheral): nu = vol-of-vol, xiCurve = forward variance (= theta),
#   al = H + 1/2, lam = lambda. Prices use S0=1 normalized OTM then scale by S0.
args <- commandArgs(trailingOnly=TRUE)
S0    <- as.numeric(args[1])
K     <- as.numeric(args[2])
T     <- as.numeric(args[3])
H     <- as.numeric(args[4])
lam   <- as.numeric(args[5])
theta <- as.numeric(args[6])   # long-run variance -> xiCurve constant
nu    <- as.numeric(args[7])   # vol-of-vol
rho   <- as.numeric(args[8])
r     <- as.numeric(args[9])

source("roughHestonAdamsLambda.R")
source("BlackFormula.R")

params <- list(al=H+0.5, lam=lam, nu=nu, rho=rho)
xiCurve <- function(t){ rep(theta, length(t)) }

nSteps <- 252
k <- log(K / S0)

otm <- function(kk){ S0 * otmRoughHeston.Adams(params, xiCurve, nSteps)(kk, T) }

# otm(k) returns the out-of-the-money option price at log-moneyness k.
# CALL: if K >= S0 (k >= 0) it is the OTM call; else use put-call parity (r=0): call - put = S0 - K
# PUT:  if K <= S0 (k <= 0) it is the OTM put;  else use parity.
if (k >= 0) {
  call_price <- otm(k)
  put_price  <- call_price - (S0 - K)
} else {
  put_price  <- otm(k)
  call_price <- put_price + (S0 - K)
}

cat(sprintf("%.12f %.12f\n", call_price, put_price))

File: bin/rough_heston_breneis/price_european.py
#!/usr/bin/env python3
"""Rough Heston pricer via SimonBreneis/approximations (Fourier + fractional Riccati).

Usage:
  price_european.py S0 K T H lambda theta nu rho r

Prints: call_price put_price
"""
import sys, os
sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))
import numpy as np
from rHestonFourier import eur_call_put

S0 = float(sys.argv[1]); K = float(sys.argv[2]); T = float(sys.argv[3])
H = float(sys.argv[4]); lam = float(sys.argv[5]); theta = float(sys.argv[6])
nu = float(sys.argv[7]); rho = float(sys.argv[8]); r = float(sys.argv[9])

V_0 = theta  # initial variance = long-run variance
K_arr = np.array([K])

# Get prices (implied_vol=False, call=True)
prices = eur_call_put(S0, K_arr, lam, rho, nu, theta, V_0, T, r=r,
                       N=3, H=H, implied_vol=False, call=True, verbose=0)

if isinstance(prices, tuple):
    price, *_ = prices
else:
    price = prices

call = float(np.asarray(price).flat[0])
put = call - (S0 - K * np.exp(-r * T))

print(f"{call:.12f} {put:.12f}")
